File: backend/src/models/DailyData.js
import mongoose from "mongoose";

const dailyDataSchema = new mongoose.Schema({
    userId: { type: String, required: true, index: true },
    sleep: {
        duration: { type: Number, required: true },
        quality: { type: String, enum: ["good", "average", "poor"], required: true },
        start: { type: Date, required: true },
        end: { type: Date, required: true }
    },
    nutrition: {
        calories: { type: Number, required: true },
        protein: { type: Number, required: true },
        carbs: { type: Number, required: true },
        fat: { type: Number, required: true }
    },
    water_intake: { type: Number, required: true },
    energy_score: { type: Number, required: true },
    timestamp: { type: Date, required: true },
}, { collection: "daily_data", timestamps: true });

dailyDataSchema.index(
    { userId: 1, timestamp: -1 },
    { unique: true, partialFilterExpression: { userId: { $exists: true }, timestamp: { $exists: true } } }
);

export default mongoose.model("DailyData", dailyDataSchema);

File: agent/models/daily_data.py
from pydantic import BaseModel, field_validator
from datetime import datetime, timezone
from typing import Literal


class SleepData(BaseModel):
    duration: int  # minutes
    quality: Literal["good", "average", "poor"]
    start: datetime
    end: datetime

    @field_validator("start", "end", mode="before")
    def parse_ts(cls, v):
        if isinstance(v, datetime):
            return v
        return datetime.fromtimestamp(v / 1000, tz=timezone.utc)


class NutritionData(BaseModel):
    calories: int
    protein: int
    carbs: int
    fat: int



class daily_data(BaseModel):
    userId: str
    sleep: SleepData
    nutrition: NutritionData
    water_intake: float
    energy_score: int
    timestamp: datetime

    @field_validator("timestamp", mode="before")
    def parse_ts(cls, v):
        if isinstance(v, datetime):
            return v
        return datetime.fromtimestamp(v / 1000, tz=timezone.utc)
